### ISSUE TITLE
sync: Consider the design of syncing `instructions.md` and source data

### ISSUE THREAD

**Issue by ee7** [exercism/configlet#179]
### Background
We want `configlet` to be able to update some track repo files from `exercism/problem-specifications`:
- a practice exercise's `.docs/instructions.md` file
- a practice exercise's `.docs/introduction.md` file, if it exists 
- a practice exercise's `.meta/config.json` file with the `source` and `source_url`

See [the docs for practice exercises](https://github.com/exercism/docs/blob/main/anatomy/tracks/practice-exercises.md#files).

`configlet` already has  a `sync` command, which helps maintainers keep the test cases of practice exercises up to date, but `configlet` does not yet support syncing other data from `problem-specifications`.

 

### Some possibilities for syncing `instructions.md` and source data
1. Implement it as the `generate` command, which existed in the `configlet` for Exercism v2 for producing docs. But that makes less sense now - we will no longer "generate" something.
1. Implement it a dedicated command with a name that isn't `generate`. This could also involve renaming the current `sync` command. But conceptually this would be like having two separate commands for dealing with syncing from `problem-specifications` (essentially `configlet sync-docs` and `configlet sync-tests`, even if they're named differently) which brings us to...
1. Add it to the existing `sync` command somehow. This could feel nice conceptually, grouping the operations that `configlet` performs on practice exercises under one command. There are several possible implementations: just use `sync`; add more options; or add subcommands. To illustrate, consider a user that wants to update a track's `instructions.md` files. We could make them run:
   - a) `configlet sync`, which has the current function, and also non-interactively updates `instructions.md` files
   - b) `configlet sync`, which has the current function, and asks the user whether they want to update the `instructions.md` files. 
   - c) `configlet sync`, which asks the user what they want to do 
   - d) `configlet sync --documentation`, which only updates instructions, possibly asking for confirmation. 
   - e) `configlet sync --documentation`, which updates instructions and performs the current `sync` behaviour, possibly with interaction
   - f) `configlet sync documentation`. This is the "add subcommands" approach: we'd also add e.g. `sync canonical-data` or `sync tests`, and `sync source-data`. The default behaviour of `sync` could be `sync all`, `sync canonical-data`, or some interactive menu. We'd have to decide what `sync canonical-data source-data` should do: sync both, or produce an error?
 
With `3d`, the current `--mode` flag would be ignored if `--documentation` is passed. We'd have to decide what to do about `--check` - a maintainer might want a CI step to fail if there are pending docs updates.

With `3f`, the `--mode` option would become an option only for `sync canonical-data`, leaving `--exercise, --prob-specs-dir` and `--offline` as `sync` options. Again, we'd need to decide what `--check` should do.

The current option text for `sync` is:
```
  -e, --exercise <slug>        Only sync this exercise
  -c, --check                  Terminates with a non-zero exit code if one or more tests are missing. Doesn't update the tests
  -m, --mode <mode>            What to do with missing test cases. Allowed values: c[hoose], i[nclude], e[xclude]
  -p, --prob-specs-dir <dir>   Use this `problem-specifications` directory, rather than cloning temporarily
  -o, --offline                Do not check that the directory specified by `-p, --prob-specs-dir` is up-to-date
```

### Discussion
We should think what we want to optimize for. Is it, for example:
1. Making it easy for the user to update `tests.toml` independently from `instructions.md` and `.meta/config.json`?
2. Minimizing the minimum number of keystrokes to update everything about practice exercises?

We should think about our recommended workflow for a maintainer, and what kind of PRs that we want to encourage on tracks. For example, making `configlet sync` update everything by default would encourage PRs to be less atomic and harder to review.

### Questions
1. ~~Does the background section cover everything that we plan to sync between `problem-specifications` and practice exercises?~~
2. ~~Is `problem-specifications` the only place from which we plan to sync files?~~
3. Are updates to `instructions.md` updates generally independent from updates of test cases? We don't want tracks to only update test cases if those test cases only make sense with an updated `instructions.md`.
4. What do we want to optimize for?

**Comment by ErikSchierboom** [comment]
1. It is missing syncing of a practice exercise's .docs/introduction.md file (which might not be present). Other than that, that is correct.
2. Yes.
3. No way to tell. This will depend on the PR. I would expect the common case to be updates to test cases and instructions to be independent.

**Comment by ErikSchierboom** [comment]
4. This is much harder to determine without having input from maintainers. Personally, I'd expect the instructions to be more important than the canonical test cases.

**Comment by ee7** [comment]
> > 3. Are updates to `instructions.md` updates generally independent from updates of test cases?
> 
> No way to tell. This will depend on the PR. I would expect the common case to be updates to test cases and instructions to be independent.

Do we require that:
- every new test case for an exercise in `problem-specifications` still makes sense with that exercise's old `description.md`? 
- every update to a `description.md` makes sense with every old test case?

And if not, should we?

I think that means it would be safer for e.g. the changes from "sync the tests" to be merged independently of "sync the docs". This makes the syntax of `configlet sync --docs` and `configlet sync --tests` more reasonable. That is, if tests and docs cannot be updated independently, we could argue that there should just be a `configlet sync` command that updates both.

If we don't have those requirements, wouldn't we need to add a field to each test case in `canonical-data.json` to describe the minimum required version of `description.md`? And things become messy if we add test cases that have mutually incompatible updates to `description.md`?

**Comment by ee7** [comment]
Another question: `.docs/instructions.md` and `.docs/introduction.md` should always be updated together, correct? Otherwise we might want e.g. `configlet sync --instructions` and `configlet sync --introduction`.

(I'm not saying that `configlet sync --some-options` is my preferred syntax).

**Comment by ErikSchierboom** [comment]
> Another question: .docs/instructions.md and .docs/introduction.md should always be updated together, correct? 

Correct.

**Comment by ErikSchierboom** [comment]
> Do we require that:
> 
> every new test case for an exercise in problem-specifications still makes sense with that exercise's old description.md?
> every update to a description.md makes sense with every old test case?
>
> And if not, should we?
>
> I think that means it would be safer for e.g. the changes from "sync the tests" to be merged independently of "sync the docs". This makes the syntax of configlet sync --docs and configlet sync --tests more reasonable. That is, if tests and docs cannot be updated independently, we could argue that there should just be a configlet sync command that updates both.

I think that it is totally reasonable for tests and docs to be updated independently. They are closely linked to each other but in the general case, one should be able to update them individually without issue.

**Comment by coriolinus** [comment]
> Some possibilities for syncing `instructions.md` and source data

IMO 3.f is the clear winner among those options: it's explicit, and allows putting flags like `--mode` in a place where they're unambiguous.

I have a preference that CLI tools are only interactive when explicitly requested to be.

> With 3f... again, we'd need to decide what `--check` should do.

It seems clear to me that in this case, `--check` returns non-0 if any unapplied updates are available. I.e. we'd have a return code table like:

Command \ Updates in: | Canonical Data | Documentation | Return code | Notes
:------------------------ | :--------------: | :--------------: | :--------: | :---------
`configlet sync --check` | true | false | nonzero | defaults to `all` behavior if subcommand not specified
`configlet sync --check` | false | true | nonzero | 
`configlet sync --check` | true | true | nonzero | 
`configlet sync docs --check` | irrelevant | true | nonzero | `docs` command is nonzero only for pending docs updates
`configlet sync tests --check` | true | irrelevant | nonzero |  `tests` command is nonzero only for pending tests updates
-- | -- | -- | 0 | all other cases return 0

> We should think what we want to optimize for.

I like granular PRs and wide-ranging checks. If it were up to me, I'd design the cli such that if you just run `sync`, it defaults to checking everything. If you do `sync <slug>`, it updates only that exercise. You can customize that behavior with `--all`, `--check`, `--update` flags. It's kind of a pain to implement, because the default behaviors are contextual, but when using it, it just does the right thing all the time. 

I want to | simple form | full form
---- | ---- | ----
Check all exercises | `configlet sync` | `configlet sync --all --check`
Update one exercise | `configlet sync <slug>` | `configlet sync <slug> --update`
Check one exercise | `configlet sync <slug> --check` | `configlet sync <slug> --check`
Update all exercises | `configlet sync --all --update` | `configlet sync --all --update`

Naming dogshedding welcome, of course; in particular, the way I wrote it, there is likely to be confusion between the `--all` flag and the `all` subcommand. Maybe the subcommand gets named `both` or something?

**Comment by ErikSchierboom** [comment]
> If it were up to me, I'd design the cli such that if you just run sync, it defaults to checking everything. 

Same here.

> If you do sync <slug>, it updates only that exercise

I like this idea of the slug being the sole parameter (without requiring a flag), as that will be the most used filtering option. `configlet sync <slug>` would slightly conflict with `configlet sync docs <slug>`, as configlet would have to know whether to interpret the first argument as the slug or as the command. We could detect this, but it does mean that not all slug names will be valid for exercises anymore, unless we keep supporting the `-e/--exercise` argument too.

Whty would we need the `--all` argument at all, if checking everything is the default anyway? Same thing for `--update`, which it being the default could mean we don't need to have an argument for it.

**Comment by coriolinus** [comment]
> `configlet sync <slug>` would slightly conflict with `configlet sync docs <slug>`

Oh, good catch. Agree that the two options there are to make some strings invalid as exercise slugs, or to support a `-e/--exercise` argument as a fallback. I have a slight preference for just saying that neither `docs` nor `tests`, nor anything else we'd use as a subcommand here, are valid exercise slugs. We'd have to add some code enforcing that in the lint portion. 

> Why would we need the `--all` argument at all... Same thing for `--update`

So that it's possible to update all exercise simultaneously. I wouldn't want to do that, but it feels like a capability that might be expected. That said, as each of those flags are only required together, for one case, I suppose it wouldn't hurt to combine them into one `--update-all` flag.

**Comment by ErikSchierboom** [comment]
> So that it's possible to update all exercise simultaneously.

But wouldn't `configlet sync --update` do that? As in: no slug was specified, hence all exercises are updated?

**Comment by coriolinus** [comment]
> But wouldn't `configlet sync --update` do that?

I think we've kind of converged on the same design, but with different names. I think `--update-all` is a bit more explicit, and therefore better; the extra chars don't matter as long as we provide a tab completion for the tool. Don't feel super strongly about the name though.

I do like getting rid of the `--all` flag, as it makes the `all` subcommand unambiguous.

**Comment by SleeplessByte** [comment]
I agree with Peter. 3 is for me the best solution, 2 is okay too.

I would really like it for `generate` to still exist, which would run inserts or whatever other logic there is, if there is any, such as inserts, but that's a nice to have.

**Comment by ErikSchierboom** [comment]
> I would really like it for generate to still exist, which would run inserts or whatever other logic there is, if there is any, such as inserts, but that's a nice to have.

We currently planned on doing these inserts _just-in-time_, as in: when the CLI requests them or when the editor displays them.

I assume the reason you would want the generate is that you'd like the see the end result?

**Comment by ee7** [comment]
I'd like to take a quick step back and ask (possibly stupid) questions about the motivation. I didn't see the answers documented [here](https://github.com/exercism/docs/blob/main/anatomy/tracks/practice-exercises.md) - if it's discussed elsewhere, please point me there.

If a practice exercise with canonical data has a `.docs/instructions.md` file that is just a copy of `description.md` from `problem-specifications`:
1) Are we sure that we want to keep a copy in every track repo? Why is it better than serving the `description.md` from prob-specs for practice exercises that have canonical data (or "practice exercises that have canonical data and are missing `instructions.md`")?
2) And if we are sure about the copy, are we sure that `configlet` should handle updating it? Why is it better than a scripted PR to every relevant track when a prob-specs `description.md` is updated, or a weekly/fortnightly/monthly PR that performs "update all outdated instructions"? I think we already planned to open/update issues on relevant tracks when test cases are changed in prob-specs.

I can see the below reasons to keep a copy. Are there more?
- It transfers complexity from the frontend to the backend: it makes every practice exercise directory look the same, even though not all practice exercises are from prob-specs.
- When a student starts an exercise, we lock everything so that the exercise doesn't change under their feet. However, it seems easy to lock the file served from prob-specs too.
- It allows a track to consciously decide that an upstream change to `description.md` is not desired. But I think the intention is for tracks to not alter `instructions.md` themselves, so it seems like deliberately not updating is not encouraged.
- It allows a track to have a practice exercise with the same name as an exercise in problem-specifications, but implementing non-canonical test cases that do not fit the prob-specs `description.md`. However, I don't see how `configlet` can behave correctly in that scenario anyway.

And I can see the below reasons to **not** keep a copy. Are there more?
- It saves maintainers time. It's true that it's not that difficult for maintainers to run some `configlet` command to update the practice exercise docs from time to time - it just feels like the new version of `configlet generate`. However, the justification for `generate` is more obvious: the files generated files are at least unique to the track. Syncing these files on every track and over the upcoming years will add up to a large amount of time - I just want to check that time investment is necessary, and can't be spent developing tracks in ways that feel more meaningful.
- `configlet` benefits from the simplicity - the `sync` command can act only on test cases and stay mostly as-is.
- It reduces noise in a track's git log.

I'm aware of https://github.com/exercism/docs/blob/main/contributing/chestertons-fence.md as I write this. But in this case I think I'm asking "why are we planning to build that fence?", rather than "why is that fence there?". And I'm more deeply involved than most; if I'm wondering about it then other people probably are too, and it's good to document the reasons in this issue.

**Comment by iHiD** [comment]
@ee7

To quote myself from [the PR that intro'd this](https://github.com/exercism/docs/pull/36#issuecomment-772756028):

> The fundamental reason is that track repos are "live", whereas problem-specs is not. When a track repo is updated the website syncs it immediately. It is effectively automatically deployed. Problem-specs is instead a reference repo. Tracks may be out of date with it (in fact, most tracks are out of date with prob-specs). Further, it is important that tracks have the ability to keep their content out of sync with problem-specs until they deliberately choose to update, else their tests etc might become out of sync with the docs in problem-specs.

> And I'm more deeply involved than most; if I'm wondering about it then other people probably are too, and it's good to document the reasons in this issue.

Yep. Wise. We are trying to do that in these PRs etc. When we get the `product` docs done that explain why things fit together, I'll copy this into there then.  My focus atm is on getting us to the "finishing line" of v3 though, so things are a little all-over-the place :)

**Comment by cmccandless** [comment]
I think @coriolinus' breakdown (https://github.com/exercism/configlet/issues/179#issuecomment-777803777) makes sense.

As far as the discussion on motivation; As a track maintainer, I'm all for almost anything that reduces noise, complexity, or redundancy in the track repos. That said, I do track with @ee7’s list of advantages.

I don't really have a personal stake in whether configlet is the tool that does this, but I think that auto-syncing description.md is a good idea.

**Comment by iHiD** [comment]
1. I'm set on having copies of things (as per my comment above). It's not a can of worms I want to reopen :)
2. We therefore need some syncing tool. It seems logical to me that we give maintainers one too to do everything with (ie `configlet` rather them needing different tools for different things. In my eyes, `configlet` is **the** maintaining tool of Exercism.

> Syncing these files on every track and over the upcoming years will add up to a large amount of time - I just want to check that time investment is necessary, and can't be spent developing tracks in ways that feel more meaningful.

I think it would make sense for tracks to turn this on as a weekly (or whatever) job, that runs `configlet generate` (or whatever other configlet commands are good) and auto PR the changes. Then the maintainer can scan and merge them. Some tracks may want this and others may not, but it feels like a low-hanging way for tracks to stay in sync with problem-specifications with minimal work. We could create an `action` for this for tracks to use.

**Comment by SleeplessByte** [comment]
(just to clarify: I am 100% pro on the just-in-time inserts. But seeing what it will likely look like is really valuable for maintainers)

**Comment by SleeplessByte** [comment]
And so. 


This to me is what I do NOT want:

```bash
configlet sync docs -e slug1
configlet sync tests-e slug1
configlet sync docs -e slug2
configlet sync tests-e slug2
```

This is what I'll be okay with:

```bash
configlet sync -e slug1 -e slug2 --docs --tests
```

I'm also okay with

```bash
configlet sync slug1 slug2 --docs --tests
```

Also okay with:

```bash
configlet sync
# assuming ALL slugs
# assuming ALL syncs, such as --docs and --tests
```

**Comment by ErikSchierboom** [comment]
The two commands that I would personally would use most would be:

`configlet sync
# Sync all things for all slugs
`

`configlet sync slug
# Sync all things for one slug
`

I think we should optimize for this. Having `--docs` and `--tests` arguments seems like a nice way to limit what is happening.

**Comment by ErikSchierboom** [comment]
> (just to clarify: I am 100% pro on the just-in-time inserts. But seeing what it will likely look like is really valuable for maintainers)

Absolutely. We should add this to configlet.

**Comment by wolf99** [comment]
> if you just run sync, it defaults to checking everything. If you do sync <slug>, it updates only that exercise.

An operation that is different depending on the operand? Sorry @coriolinus but, 🤮.

As a user of a CLI tool, I would like an operation without flags to do the same type of action irrespective of the number of operands.
As a silly human, I would prefer the default commands to be non-destructive, i.e. `check`, rather than `update`.
Though if you prefer sync to be the default rather than check then so be it. 
But please don't mix and match.

That is:

I want to | simple form | full form
-- | -- | --
Check all exercises | configlet sync | configlet sync --all --check
Check one exercise | configlet sync <slug> | configlet sync <slug> --check
Update one exercise | configlet sync <slug> --update | configlet sync <slug> --update 
Update all exercises | configlet sync --all --update | configlet sync --all --update

Or:

I want to | simple form | full form
-- | -- | --
Update all exercises | configlet sync | configlet sync --all --update 
Update one exercise | configlet sync <slug> | configlet sync <slug> --update 
Check one exercise | configlet sync <slug> --check| configlet sync <slug> --check
Check all exercises | configlet sync --all --check| configlet sync --all --check

But not :

I want to | simple form | full form
-- | -- | --
Check all exercises | configlet sync | configlet sync --all --check
Update one exercise | configlet sync <slug> | configlet sync <slug> --update
Check one exercise | configlet sync <slug> --check | configlet sync <slug> --check
Update all exercises | configlet sync --all --update | configlet sync --all --update

**Comment by ErikSchierboom** [comment]
I would argue for the below commands:

Action | Command
 -- | --
Update all exercises | `configlet sync`
Update one exercise | `configlet sync <slug>`
Check all exercises | `configlet sync --check`
Check one exercise | `configlet sync <slug> --check`

The defaults are: 

- Do an update instead of a check: I feel that this is how people would use the sync command most of the time
- Operate on all exercises
- Update all types of documents (docs and tests at the moment)

And then to allow syncing only specific things, we can have arguments:

- `--docs` for the documentation
- `--tests` for the tests.toml files

Action | Command
 -- | --
Update only the docs of all exercises | `configlet sync --docs`
Check only the tests of one exercise | `configlet sync <slug> --tests --check`

**Comment by ee7** [comment]
I'm mostly aligned with the above comment (although I might argue later for `configlet sync -e slug` over `configlet sync slug`. 

I think `configlet sync` will be the most used `sync` command, and so one of the things I want to optimize for is the usefulness of its default behavior. Therefore I prefer `configlet sync` to `configlet sync --update` (if `sync` doesn't do anything unless `--update` is passed then maybe `sync` is a bad name).

But to address @wolf99's comment of
> I would prefer the default commands to be non-destructive, i.e. `check`, rather than `update`.

I think we can get the best of both worlds - we can have a menu at the start of `configlet sync` so that it doesn't immediately prompt the user for y/n choices.

Some advantages:
- We can print a summary of the track's practice exercise status, which is even more useful now that `sync` will be handling both docs and tests
- It removes some of the need to run `configlet sync --check` separately first. The `--check` option can be more for the exit code, rather than "find a slug to pass as an argument".
- It allows updating exercises in a non-alphabetical order without passing a slug as an argument. 

So in the long-term I think we could implement something like:

```
$ ./configlet sync

Available updates to practice exercises on the Ruby track:

| exercise    | number of new tests | outdated docs |
|-------------+---------------------+---------------|
| bob         |                   7 |               |
| hello-world |                     | yes           |
| two-fer     |                   2 | yes           |

Available operations:
- q    quit
- t    Include or exclude missing tests
- d    Update docs (including source information)
- a    Update all tests and all docs
- v    Show full details of missing tests
- 1    bob: update tests
- 2    hello-world: update docs
- 3    two-fer: update tests
- 4    two-fer: update docs

Your choice: 
```

Some questions that apply regardless of whether we do the above:
1. Should we ask the user about doc changes, or do they happen automatically?
1. If we have `--docs` and `--tests`, should we remove the `-m, --mode choose|include|exclude` option and support `--tests choose|include|exclude`?
1. If the user wants to update docs and tests, in what order should `configlet sync` do it? Let's say there are updates to: the tests of `bob`, the docs for `hello-world`, and the docs+tests for `two-fer`. Should we update all docs first, or update both docs and tests alphabetically by exercise? Currently, we write to an exercise's `tests.toml` when the user finishes making choices for that exercise; a keyboard interrupt in the middle of an exercise only discards the choices related to that exercise.  
1. Should `configlet sync --docs` involve interaction?
1. Should `configlet sync bob --docs` involve interaction?
1. If yes to the above two, what should be the syntax for interaction-free updating of docs?

**Comment by coriolinus** [comment]
I don't mind an interactive mode, but to my mind, interactive mode is an additional feature; the fundamentals should be expressed via non-interactive CLI options. It's _much_ easier to script against a non-interactive CLI. What that in turn means is that I would strongly prefer that the interactive TUI is both accessed via an explicit flag, and is not the first priority for development.

**Comment by iHiD** [comment]
I don't have opinions on most of this, but I do agree that from a priority POV, we should get everything working non-interactively first, then do the interactivity later. For now, tracks are waiting on the ability for configlet to do things en-mass for all the various changes that are happening, so it seems to me that the crucial functionality is that they can get stuff done, rather than being able to do it "nicely". After all, any changes can be checked during the PR phase, and manual changes akin to the interactivity applied then.

**Comment by ErikSchierboom** [comment]
I personally don't really see me using the interactive mode. I'd much rather (like @coriolinus), make it simple for the CLI to be non-interactive. I'd thus suggest to use the commands I proposed in https://github.com/exercism/configlet/issues/179#issuecomment-779698733 and we could potentially revisit an interactive mode sometime in the (far) future.

**Comment by ErikSchierboom** [comment]
One new sync option to add: syncing the `files` property in the `.meta/config.json` file, provided the track has defined the `files` pattern in its `config.json`. I would propose to use a `--filepaths` argument for that.

**Comment by ee7** [comment]
I think I'm missing something.

Facts:
- `sync` currently handles only `tests.toml`
- `configlet sync` is currently interactive

My understanding:
- the `sync` command overall will soon also handle syncing docs
- syncing docs is mostly safe to do with only a minimal review, but syncing `tests.toml` generally requires a maintainer's brain in the loop
- a bare `configlet sync` will "sync all things for all slugs"
- therefore a bare `configlet sync` must involve interaction, as it does now

---

@ErikSchierboom, given
> I personally don't really see me using the interactive mode. 

What happens when a user runs a bare `configlet sync` if there is no interaction?

**Comment by ErikSchierboom** [comment]
>> I personally don't really see me using the interactive mode.
> What happens when a user runs a bare configlet sync if there is no interaction?

What I meant was the extended interactive mode that was suggested, I'm fine with what we already have. That said, there is a dilemma here regarding sync updating everything and requiring interactivity.

Choosing which tests to include/exclude in the tests.toml files, that _does_ indeed require a maintainer's brain. I'm not entirely sure what the best approach would be. Some options:

- A bare configlet sync will have interaction for the tests.toml files
- A bare configlet sync will include/exclude/skip all tests (no interaction needed)
- A bare configlet sync will output any missing tests but not do anything (skipping) (no interaction needed)
- A bare configlet sync will not do any updates and only output what requires updating (no interaction needed). Updating would require a flag.

@SleeplessByte @wolf99 @coriolinus What do you think?

**Comment by coriolinus** [comment]
My preference is that bare `configlet sync` does a check, not an update, in all cases; that corresponds to option 4 above. It's simple and consistent. You can then do `configlet sync --update --toml --include` or whatever when you want update actions.

**Comment by SleeplessByte** [comment]
My preference is that everything that REQUIRES interaction errors by default OR that it just provides the interaction but there is a `--no-interaction` flag. I think the consensus in this thread is that we prefer `--interactive / -i` instead. I don't have an opinion on that.

(So this is op top of @coriolinus comments, not instead of)

**Comment by ee7** [comment]
@coriolinus wrote
> I don't mind an interactive mode, but to my mind, interactive mode is an additional feature; the fundamentals should be expressed via non-interactive CLI options. It's much easier to script against a non-interactive CLI.

Usually, I agree. But in this case I was leaning towards interactivity by default because:
- Updating `tests.toml` fundamentally involves decisions by a maintainer - tracks shouldn't automatically add all new tests to `tests.toml`
- The `sync` command will handle updating `tests.toml`
- The other functions of `sync` seem more trivial and less important to optimize for
- It seems time-efficient to present the user with choices about unseen tests
- I'd like for `configlet sync` to address the most common use cases without requiring a large number of flags

But I'm happy to be swayed: @coriolinus can you describe what kind of scripting you want to do, and what kind of workflow you envisage?

Below are some workflows that seem reasonable to me. Does anybody plan on doing something else?
- The maintainer watches the problem-specifications repo and runs some `configlet` command afterwards
- The maintainer runs some `configlet` command locally whenever convenient and updates things as they desire
- More ambitious automating:
   - A change in `problem-specifications` to an exercise's docs automatically creates a corresponding PR on a track. The maintainer reviews this PR, but usually no changes are required.
   - A change in `problem-specifications` to an exercise's `canonical-data.json` automatically creates a draft PR on a track to includes new test cases in `tests.toml` and the exercise's test file (re-running a track's test generator). The maintainer reviews the PR carefully, and excludes tests sometimes.

We can support some syntax like
```
configlet sync --include-test f3f85018-9b54-4f13-aab4-b51743895767
```

but the script that calls that already needs to know that it's sensible to include that UUID - I don't see how it fits in.

**Comment by ErikSchierboom** [comment]
> The maintainer watches the problem-specifications repo and runs some configlet command afterwards
> The maintainer runs some configlet command locally whenever convenient and updates things as they desire

I think these two workflows are the most common. The automation could be useful, but I wouldn't want to optimize for that as a) it will probably be relatively rare and b) automation will be scripted so it can be slightly less convenient.

> My preference is that everything that REQUIRES interaction errors by default OR that it just provides the interaction but there is a --no-interaction flag. I think the consensus in this thread is that we prefer --interactive / -i instead. I don't have an opinion on that.

I think this appears to be the general consensus, and I'm fine with it being non-interactive by default.

**Comment by cmccandless** [comment]
> My preference is that bare configlet sync does a check, not an update, in all cases; that corresponds to option 4 above. It's simple and consistent. You can then do configlet sync --update --toml --include or whatever when you want update actions.

I'm of this mind as well.

> and I'm fine with it being non-interactive by default.

Yes.

**Comment by coriolinus** [comment]
> can you describe what kind of scripting you want to do, and what kind of workflow you envisage?

I don't want to watch _anything_ manually. I expect to set up a CI job which checks for updates on some schedule, possibly weekly. For each exercise with non-`tests.toml` updates, it'll create a single PR. Creating the PR notifies me, at which point if all is well, all I have to do is approve and merge.

Updates to `tests.toml` are a bit harder, because as you say, those require more intervention. I expect it to create a PR for each exercise with `tests.toml` updates. By default, it'll include all updates in that file, but then fail CI somehow. (Haven't designed that part thoroughly yet in my head.) CI passes when `tests.toml` once again matches the exercise's tests, at which point we can merge.

I'm generally happy to spend some effort up front writing scripts, because it makes it much more likely that updates over time will happen in a reasonably efficient way.

**Comment by SaschaMann** [comment]
I've only skimmed all the comments, but Erik asked me to chime in. Some of what I say may have been mentioned already but I missed it, sorry. I don't have any opinions on the CLI design.

For me, this is the workflow I'd want configlet to be optimised for:

> A change in problem-specifications to an exercise's docs automatically creates a corresponding PR on a track. The maintainer reviews this PR, but usually no changes are required.

> A change in problem-specifications to an exercise's canonical-data.json automatically creates a draft PR on a track to includes new test cases in tests.toml and the exercise's test file (re-running a track's test generator). The maintainer reviews the PR carefully, and excludes tests sometimes.

A merged PR in prob-specs triggers a gha workflow in the track repo that then generates a PR.

> I think these two workflows are the most common. The automation could be useful, but I wouldn't want to optimize for that as a) it will probably be relatively rare and b) automation will be scripted so it can be slightly less convenient.

I don't think this assumption can be made. Currently these manual workflows are the most common because there aren't good alternatives but with [probby](https://github.com/exercism/probby) this may change. Currently problem-specs takes less priority compared to writing enough concept exercises, so I haven't spent time on probby, but in the long run I think a non-manual/push-to-prob-specs-based workflow has a good chance of becoming the most common workflow. Optimising in favour of manual syncing/against automation seems premature.

---

This also feels like a feature that could wait for a few months because practice exercise don't take priority for most tracks right now, and prob-specs is fairly silent.

**Comment by ErikSchierboom** [comment]
> A merged PR in prob-specs triggers a gha workflow in the track repo that then generates a PR.

I think not every track would want this. But 🤷 

> Optimising in favour of manual syncing/against automation seems premature.

Possibly, but I don't think the suggestions make automation significantly harder, just slightly more verbose.

> This also feels like a feature that could wait for a few months because practice exercise don't take priority for most tracks right now, and prob-specs is fairly silent.

I disagree, for the following reasons:

- The design discussed here is important not only for this specific command (syncing), but also as a general design for other commands. We need to implement this pattern before we release our first stable version to prevent breaking changes.
- Many tracks will likely not have Concept Exercises for a while, so they'll solely be working with Practice Exercises
- Prob-specs is indeed relatively silent, but there _are_ still things happening there. 
- I expect tracks to have outdated documents so we should make it easy for tracks to sync to the latest version.

**Comment by SaschaMann** [comment]
> Possibly, but I don't think the suggestions make automation significantly harder, just slightly more verbose.

To clarify: I wasn't implying that. I just outlined the workflow I'd want to use. I don't care too much about the details of the CLI design.

**Comment by ErikSchierboom** [comment]
Taking into account all the suggestion, the final spec is:

## Spec

Action | Command | Short command
 -- | -- | --
Check all exercises | `configlet sync ` | `configlet sync `
Check one exercise | `configlet sync --exercise <slug>` | `configlet sync -e <slug>`
Update all exercises | `configlet sync --update` | `configlet sync -u`
Update one exercise | `configlet sync --exercise <slug> --update` | `configlet sync -e <slug> -u`

## Slug as flags vs positional argument

For now, we'll be passing the exercise slug explicitly via an argument (`-e` or `--exercise`). There are pros and cons of positional arguments vs flag arguments, see https://clig.dev/#arguments-and-flags and https://news.ycombinator.com/item?id=25304257 for a discussion of those. We might want to support the positional argument approach later, but for now we'll just use the existing flag-based option.

## Default

The defaults are: 

- Do a check instead of an update (non-destructive)
- Operate on all exercises
- Sync will sync all types of documents (docs and tests at the moment)

## Flags

And then to allow syncing only specific things, we can have the following flags:

- `--docs`: for the `.docs/introduction.md` and `.docs/instructions.md` files of practice exercises
- `--metadata`: for the `source`, `source_url` and `blurb` properties in the `.meta/config.json` files of practice exercises
- `--tests`: for the `tests.toml` files of practice exercises
- `--filepaths` for the `files` entries in the `.meta/config.json` files (practice _and_ concept exercises)
- `--exercise <slug>`: only sync this exercise

**Comment by ee7** [comment]
> For me, this is the workflow I'd want configlet to be optimised for:
> 
> > A change in problem-specifications to an exercise's docs automatically creates a corresponding PR on a track. The maintainer reviews this PR, but usually no changes are required.
> 
> > A change in problem-specifications to an exercise's canonical-data.json automatically creates a draft PR on a track to includes new test cases in tests.toml and the exercise's test file (re-running a track's test generator). The maintainer reviews the PR carefully, and excludes tests sometimes.

This is pretty much the workflow that I'd want as a maintainer. I just didn't know whether configlet should optimize for that flow, given that we've never had it, but I'll agree with both of:

> > Optimising in favour of manual syncing/against automation seems premature.
> 
> Possibly, but I don't think the suggestions make automation significantly harder, just slightly more verbose.

**Comment by ee7** [comment]
The specification in https://github.com/exercism/configlet/issues/179#issuecomment-788842323 doesn't say anything about interactivity. We need to solidify that aspect.

I agree with the previous points that an interactive mode should not be the first priority for development. However, I don't see a design that fits an explicit `--interactive` flag (and the best design with interaction as opt-out isn't obvious either).

Note that even if interactivity is the default, we can still add the non-interactive parts to `configlet` first. There'd just be a period where we print a usage message instead of prompting.

I'll write a series of points below - they're numbered just so it's easier to refer to them individually. They aren't mutually exclusive alternatives. I'll try to argue from first principles and summarize some previous discussion.

Any thoughts welcome.


### 1. Is `configlet sync --update` interactive?
1a. The spec says `sync --update` will operate on everything (tests, docs, metadata, and probably filepaths) by default.

1b. Syncing tests requires a human making a decision. Including every new test case should not be the default.

1c. By 1a and 1b, `sync --update` shouldn't non-interactively update a `tests.toml`.

1d. Therefore `sync --update` must either do nothing (printing a usage message and saying basically "use more options"), or be interactive.

1e. If it is interactive, it should operate on docs/metadata/filepaths before tests. That's because we should continue to support a keyboard interrupt properly - currently, configlet writes `tests.toml` at the moment the user has made every include/exclude/skip decision for one exercise. If tests were handled first, then the user would only get prompted about docs/metadata/filepaths when there are no more test case decisions. And test case decisions are harder.


### 2. Are `configlet sync -u --tests` and `configlet sync -u --docs` interactive?
2a. For consistency, the interactivity of `sync --update --tests` should match that of `sync --update`.

2b. Therefore if `sync --update` does nothing but print a usage message, so should `sync --update --tests` (and we insist on `sync --update --tests --mode choose` or e.g. `sync --update --tests --interactive`).

2c. Similarly, if `sync --update` does nothing but print a usage message, so should `sync --update --docs` (and we insist on e.g. `sync --update --docs -y` or `sync --update --docs --interactive`).

2d. It seems unfriendly for every command to do nothing but print a usage message unless e.g. `-y` or `--interactive` is passed. And `sync --update --docs -y` makes sense only if `sync --update --docs` actually prompts the user.

2e. I don't see a design for an `--interactive` flag that fits with (or replaces) `--mode choose` or `--tests (choose|)include|exclude`. It's unclear what `sync --tests include --interactive` would do.

2f. So overall, I don't see a design that has interaction as an explicit opt-in.

2g. People might prefer confirmation or an explicit opt-in before writing changes to a track directory.


### 3. If interaction is the default, how do we disable interactivity?

3a. If we have interaction as the default, there should be a way to skip the prompt (especially important for scripting).

3b. Something like `sync --update --docs -y` (long-form maybe `--yes` or `--accept) seems more obvious than `sync --update --docs --non-interactive` (or `--no-interaction`).

3c. A `-y` flag would clearly apply to `--docs`, `--metadata`, `--filepaths`.

3d. Given 3c, it feels messy to still have `--mode choose|include|exclude`, as a separate way to skip interaction that applies only to `--tests`. We also previously phrased the prompt like "Include this test? y/n/s", so it's confusing if `sync --tests -y` doesn't include every test.

3e. We could remove `--mode`, and change `--mode choose|include|exclude` to `--tests choose|include|exclude`. Perhaps with `choose` as a default value. But then it's strange that `--docs/--metadata/--filepaths` have a different model: `--docs choose` doesn't work (even though `--docs include` doesn't really make sense). And something like `sync --tests choose -y` is still confusing.

3f. If interaction is the default, a bare `sync --update` or `sync --update --docs` should produce an error message when stdout/stderr is not an interactive terminal, so that we insist on e.g. `sync --update --docs -y` or e.g. `sync --update --tests include` in scripts.

**Comment by ErikSchierboom** [comment]
I think it makes sense to have the docs/metadata/filepaths/tests syncing all use the same interactivity by default. As we want to encourage people syncing and syncing tests is interactive in the majority of cases, I think we should have the docs/metadata/filepaths also be interactive by default, with just a simple prompt like: "Exercises X, Y, Z have outdated metadata. Do you want to update?". As an alternative, we could prompt for each exercise where docs/metadata/filepaths need syncing, just like with the tests. This could become cumbersome, but we could change the allowed responses from "yes/no" to "yes/no/yes to all/no to all" or something similar. That would make the behavior be quite consistent along the different options.

The above suggestion makes it slightly more involved to update the docs/metadata/filepaths, so we should indeed also support disabling interactivity. Renaming `mode` to something else makes sense, as the name does not make it clear that it is linked to tests syncing. I like the idea of a `-y` option, as it is used by so many existing tools that people might intuitively try this. This works well for docs/metadata/filepaths, but would still need additional input for tests. 

What if doing `sync --update -y` or ``sync --update --tests -y` outputs an error message, like "To non-interactively sync the tests, please provide a default value using ...."

I'm not sure if allowing the `--tests` option to be passed a parameter is the best option, here. I would expect that `sync --update --tests` would (interactively) sync the tests, in which case I wouldn't need to provide a default value. Maybe a separate option is better?

Side note: we could consider having a `skip` mode as the default choice  for tests, which would allow `sync update -y --mode skip` to update everything _but_ the tests.

**Comment by ee7** [comment]
> I'm not sure if allowing the `--tests` option to be passed a parameter is the best option, here. I would expect that `sync --update --tests` would (interactively) sync the tests, in which case I wouldn't need to provide a default value. Maybe a separate option is better?

With point 3e above, I meant that
```
sync --update --tests
```

could do the same as
```
sync --update --tests choose
```

Does that change your opinion?

I'm thinking that

```
sync --update --tests include
```

is cleaner than

```
sync --update --tests --test-mode include
```

**Comment by ee7** [comment]
> What if doing `sync --update -y` or `sync --update --tests -y` outputs an error message, like "To non-interactively sync the tests, please provide a default value using ...."
> 
> Side note: we could consider having a `skip` mode as the default choice for tests, which would allow `sync update -y --mode skip` to update everything _but_ the tests.

I think I'm okay with either of these. It just depends whether we want `sync -uy` to produce an error, or to non-interactively update everything apart from tests (with an explanatory message).

**Comment by ErikSchierboom** [comment]
> With point 3e above, I meant that
> ```
> sync --update --tests
> ```
> 
> could do the same as
> ```
> sync --update --tests choose
> ```
> 
> Does that change your opinion?

Yeah that makes sense then.

> I think I'm okay with either of these. It just depends whether we want sync -uy to produce an error, or to non-interactively update everything apart from tests (with an explanatory message).

I honestly don't know the best option. I would lean towards producing an error, which can then prompt an explanatory message indicating that you can use `--tests skip` if you want to ignore tests.

**Comment by ee7** [comment]
> > I think I'm okay with either of these. It just depends whether we want `sync -uy` to produce an error, or to non-interactively update everything apart from tests (with an explanatory message).
> 
> I honestly don't know the best option. I would lean towards producing an error, which can then prompt an explanatory message indicating that you can use `--tests skip` if you want to ignore tests.

Let's go with "`sync -uy` produces an error" for now. Then if we want to change it later, it's a backwards-compatible change.

**Comment by ErikSchierboom** [comment]
I like it!

**Comment by ee7** [comment]
Something I forgot to ask: any thoughts on something like
```
--scope docs
--scope docs,tests
```
instead of `--docs` etc?

I think one problem is that then we'd still need the `--tests-mode` option, which I don't love. I've been implementing the `--docs` way, but I'll mention this in case you can see something better.

**Comment by ErikSchierboom** [comment]
I'm not a fan of the comma-separated approach to be honest. I find that less intuitive than having flags.

**Comment by ee7** [comment]
> I'm not a fan of the comma-separated approach to be honest. I find that less intuitive than having flags.

As discussed privately:
Fine by me. I think comma-separated option values is fairly uncommon, and we haven't used it elsewhere in the configlet CLI.
